RecoLocalTracker/SiPixelClusterizer/plugins/alpaka/ClusterChargeCut.h

```diff
@@ -0,0 +1,207 @@
+
+          // #ifdef GPU_DEBUG
+          // printf("module %d -> chargeCut = %d; cluster %d; charge = %d; ok = %s\n",thisModuleId, chargeCut,i,charge[i],ok[i] > 0 ? " -> good" : "-> cut");
+          // #endif
+        });
+        alpaka::syncBlockThreads(acc);
+
+        // if all clusters above threshold do nothing
+        // if (allGood)
+        //   continue;
+
+        // renumber
+        auto& ws = alpaka::declareSharedVar<uint16_t[32], __COUNTER__>(acc);
+        constexpr uint32_t maxThreads = 1024;
+        auto minClust = std::min(nclus, maxThreads);
+
+        cms::alpakatools::blockPrefixScan(acc, newclusId, minClust, ws);
+
+        if constexpr (maxNumClustersPerModules > maxThreads)  //only if needed
+        {
+          for (uint32_t offset = maxThreads; offset < nclus; offset += maxThreads) {
+            cms::alpakatools::blockPrefixScan(acc, newclusId + offset, nclus - offset, ws);
+
+            cms::alpakatools::for_each_element_in_block_strided(acc, nclus - offset, [&](uint32_t i) {
+              uint32_t prevBlockEnd = ((i + offset / maxThreads) * maxThreads) - 1;
+              newclusId[i] += newclusId[prevBlockEnd];
+            });
+            alpaka::syncBlockThreads(acc);
+          }
+        }
+
+        ALPAKA_ASSERT_OFFLOAD(nclus >= newclusId[nclus - 1]);
+
+        if (nclus == newclusId[nclus - 1])
+          return;
+
+        clus_view[thisModuleId].clusInModule() = newclusId[nclus - 1];
+        alpaka::syncBlockThreads(acc);
+
+#ifdef GPU_DEBUG
+        if (thisModuleId % 100 == 1)
+          if (cms::alpakatools::once_per_block(acc))
+            printf("module %d -> chargeCut = %d; nclus (pre cut) = %d; nclus (after cut) = %d\n",
+                   thisModuleId,
+                   chargeCut,
+                   nclus,
+                   clus_view[thisModuleId].clusInModule());
+#endif
+        // mark bad cluster again
+        cms::alpakatools::for_each_element_in_block_strided(acc, nclus, [&](uint32_t i) {
+          if (0 == ok[i])
+            newclusId[i] = invalidModuleId + 1;
+        });
+
+        alpaka::syncBlockThreads(acc);
+
+        // reassign id
+        firstElementIdx = firstElementIdxNoStride;
+        endElementIdx = endElementIdxNoStride;
+        for (uint32_t i = firstElementIdx; i < numElements; ++i) {
+          if (not cms::alpakatools::next_valid_element_index_strided(
+                  i, firstElementIdx, endElementIdx, blockDimension, numElements))
+            break;
+          if (digi_view[i].moduleId() == invalidModuleId)
+            continue;  // not valid
+          if (digi_view[i].moduleId() != thisModuleId)
+            break;  // end of module
+          if (0 == ok[digi_view[i].clus()])
+            digi_view[i].moduleId() = digi_view[i].clus() = invalidModuleId;
+          else
+            digi_view[i].clus() = newclusId[digi_view[i].clus()] - 1;
+          // digi_view[i].clus() = newclusId[digi_view[i].clus()] - 1;
```

Comment:
Would be nice to clean up commented code at some point.  It looks like a duplicated line, unless I'm missing the differenence.

Reply:
Thanks, I'll add it to the TODO list.